### PR TITLE
demos: tidy up makefiles, fix warnings

### PR DESCRIPTION
**Pull request by jamuir**
Update makefiles so that consistent patterns are used.  Object files are compiled from source using an implicit rule (but using our CFLAGS); for linking, we give an explicit rule.  Ensure that "make test" works in each subdirectory (even if it does not actually run any applications).  The top-level demo makefile now works.

The makefiles are not make-agnostic.  e.g. they use the variable $(RM) in "clean" recipes, which is defined in gnu-make but may not be defined in others.

Part of #17806 (eventually, we want the demos to be built and executed as part of our CI)

Testing:
```
$ cd demo
$ make test
```
##### Checklist

- [x ] documentation is added or updated

(demos are considered part of the documentation)

### CODE REVIEW

File: demos/kdf/hkdf.c [hunk 1]
```diff
@@ -95,6 +95,8 @@ int main(int argc, char **argv)
         goto end;
     }
 
+    printf("Success\n");
```
Comment (nhorman) [inline demos/kdf/hkdf.c:98]: Should the success message go to stdout?  or do we want it to go to stderr

Comment (paulidale) [reply demos/kdf/hkdf.c:98]: stdout seems reasonable to me.

File: demos/cms/cms_ddec.c [hunk 2]
```diff
@@ -34,7 +34,8 @@ int main(int argc, char **argv)
 
     rcert = PEM_read_bio_X509(tbio, NULL, 0, NULL);
 
-    BIO_reset(tbio);
+    if (BIO_reset(tbio))
+        goto err;
```
Comment (t8m) [inline demos/cms/cms_ddec.c:38]: This should be `if (BIO_reset(tbio) < 0)`

File: demos/cms/cms_dec.c [hunk 3]
```diff
@@ -31,7 +31,8 @@ int main(int argc, char **argv)
 
     rcert = PEM_read_bio_X509(tbio, NULL, 0, NULL);
 
-    BIO_reset(tbio);
+    if (BIO_reset(tbio))
```
Comment (t8m) [inline demos/cms/cms_dec.c:34]: As above

File: demos/cms/cms_sign.c [hunk 4]
```diff
@@ -38,7 +38,8 @@ int main(int argc, char **argv)
 
     scert = PEM_read_bio_X509(tbio, NULL, 0, NULL);
 
-    BIO_reset(tbio);
+    if (BIO_reset(tbio))
```
Comment (t8m) [inline demos/cms/cms_sign.c:41]: As above and so on below

**Review by t8m — CHANGES_REQUESTED** [review]

File: demos/kdf/hkdf.c (hunk 1 above)
Comment (jamuir) [reply demos/kdf/hkdf.c:98]: the general pattern I noticed in the demo code was that error messages go to stderr and status / success messages go to stdout, so I picked stdout above.

However, I just did `cd demos && make > stdout.txt` on my branch, and I see that `pkey/EVP_PKEY_RSA_keygen.c` does not follow that pattern (status messages go to stderr).  I will push an update for that.

File: demos/cms/cms_ddec.c (hunk 2 above)
Comment (jamuir) [reply demos/cms/cms_ddec.c:38]: okay.  I will update those checks.

I find `BIO_reset()` strange b/c its return value depends on what type of BIO is being reset:

https://www.openssl.org/docs/manmaster/man3/BIO_reset.html

> BIO_reset() normally returns 1 for success and <=0 for failure. File BIOs are an exception, they return 0 for success and -1 for failure.

Comment (t8m) [reply demos/cms/cms_ddec.c:38]: We use the boolean checks only for functions that return 0 or 1 only. As BIO reset is different (and awkward as you noticed) comparison operator must be used appropriately.

Comment (jamuir) [reply demos/cms/cms_ddec.c]: fixed.

**Review by paulidale — APPROVED** [review]

**Comment by t8m** [comment]
Merged to the master branch. Thank you.

**Comment by DDvO** [comment]
Can this, or at least the part renaming `demos/sslecho/makefile` to `demos/sslecho/Makefile`, please be backported to 3.2?
This fixes trouble switching between branches on a case-insensitive file system (such as MacOS).

**Comment by nhorman** [comment]
@DDvO that seems reasonable to me, but we will need a separate issue for that, can you open one and mark it as branch:3.2?  Bonus points if you're willing to submit the PR :)

**Pull request closed without merging**